### PR TITLE
Adding reduced GsfTracks to miniAOD

### PR DESCRIPTION
**Pull request by andreypz**
Hi,

This commit is to add reduced gsf tracks collection to miniAOD. The tracks added are:
 * The main track of GsfElectronCore
 * All ambiguous tacks of GsfElectron (if there are any)

Change in the file size was estimated with *test/patMiniAOD_standard_cfg.py* script on two RelVal samples: **TTbar** and **ZtoEE**, based on 500 events:
 * In TTbar sample the increase is from 32470 Kbytes to 32524 Kbytes (0.17%). Increase per event: 108 bytes.
 * In ZEE sample the increase from 25427 Kbytes to 25460 Kbytes (0.13%). Increase per event: 65 bytes.

This addition was presented in [X-POG meeting on 18-Oct-2016](https://indico.cern.ch/event/575788/), in [this slides](https://indico.cern.ch/event/575788/contributions/2343082/attachments/1356862/2051472/Zenz-DalitzMiniAODNeeds-XPOG-v1.pdf).  
This was also privately discussed with @arizzi @gpetruc @sethzenz @bendavid.

The intent for this code to be used for miniAOD of 2017 data/MC and legacy re-miniAOD of 2016 data.

This supersedes pull #19597

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 1]
```diff
@@ -401,6 +410,33 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
     const reco::GsfElectronCoreRef &gsfElectronCore = gsfElectron.core();
     linkCore(gsfElectronCore, *gsfElectronCores, gsfElectronCoreMap);
 
+
```
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: One empty line is enough to separate

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 2]
```diff
@@ -401,6 +410,33 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
...
+      
+      const reco::GsfTrackRef &ambigGsfTrack = *igsf;
+      if (!gsfTrackMap.count(ambigGsfTrack)) {
+	gsfTracks->push_back(*ambigGsfTrack);
+	gsfTrackMap[ambigGsfTrack] = gsfTracks->size() - 1;
+	//std::cout << "Save ambiguous track in the map" << std::endl;
+      }
+    }
+    
+
```
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: One empty line is enough to separate

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 3]
```diff
@@ -427,6 +463,14 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
     linkConversionsByTrackRef(singleConversionHandle, gsfElectron, *singleConversions, singleConversionMap);
   }
 
+
```
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: One empty line is enough to separate
Please remove the commented out lines

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 4]
```diff
@@ -568,6 +614,39 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
...
+    for (reco::GsfTrackRefVector::const_iterator igsf = gsfElectron.ambiguousGsfTracksBegin(); igsf != gsfElectron.ambiguousGsfTracksEnd(); ++igsf) {
+      const reco::GsfTrackRef &ambGsfTrack = *igsf;
+      ambigTracksInThisElectron.push_back(*igsf);
+    }
+    
+    // Now we need to clear them (they are the refs to original collection):
+    gsfElectron.clearAmbiguousGsfTracks();
+    
+    // And here we add them back, now from a new reduced collection:
+    for (std::vector<reco::GsfTrackRef>::const_iterator it = ambigTracksInThisElectron.begin(); it != ambigTracksInThisElectron.end(); ++it) {
```
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: Range based for loops must be preferred

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 5]
```diff
@@ -568,6 +614,39 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
...
+      const auto &gsftkmapped = gsfTrackMap.find(*it);
+      
+      if (gsftkmapped != gsfTrackMap.end()) {
+	reco::GsfTrackRef gsftkref(outGsfTrackHandle, gsftkmapped->second);
+	gsfElectron.addAmbiguousGsfTrack(gsftkref);
+      }
+      else
+	throw cms::Exception("There must be a problem with linking and mapping of ambiguous gsf tracks...");
+
```
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: Remove this empty line

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 6]
```diff
@@ -568,6 +614,39 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
...
+      }
+      else
+	throw cms::Exception("There must be a problem with linking and mapping of ambiguous gsf tracks...");
+      
+    }
+    
+    if (gsfElectron.ambiguousGsfTracksSize() > 0)
+      gsfElectron.setAmbiguous(true); // Set the flag
+    
+
```
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: One empty line is enough to separate

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 7]
```diff
@@ -829,6 +908,19 @@ void ReducedEGProducer::relinkSuperCluster(T& core,
...
+void ReducedEGProducer::relinkGsfTrack(reco::GsfElectronCore& gsfElectronCore, 
+				       const std::map<reco::GsfTrackRef, unsigned int>& gsfTrackMap, 
+				       const edm::OrphanHandle<reco::GsfTrackCollection>& outGsfTrackHandle) 
+{
+  const auto &gsftkmapped = gsfTrackMap.find(gsfElectronCore.gsfTrack());
+  if (gsftkmapped != gsfTrackMap.end()) {
+    reco::GsfTrackRef gsftkref(outGsfTrackHandle, gsftkmapped->second);
+    gsfElectronCore.setGsfTrack(gsftkref);
+    //std::cout<< "relinked a gsf track" << std::endl;
```
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: Please remove the commented out line

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 8]
```diff
@@ -568,6 +614,39 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
 
   for (reco::GsfElectron &gsfElectron : *gsfElectrons) {
     relinkGsfElectronCore(gsfElectron, gsfElectronCoreMap, outgsfElectronCoreHandle);
+
+    // -----
+    // Also in this loop let's reling ambiguous tracks
+    // -----
+    std::vector<reco::GsfTrackRef> ambigTracksInThisElectron;
+    // Here we loop over the ambiguous tracks and save them in a vector
+    for (reco::GsfTrackRefVector::const_iterator igsf = gsfElectron.ambiguousGsfTracksBegin(); igsf != gsfElectron.ambiguousGsfTracksEnd(); ++igsf) {
```
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: Range based for loops must be preferred

File: RecoEgamma/EgammaPhotonProducers/interface/ReducedEGProducer.h [hunk 9]
```diff
@@ -106,7 +106,12 @@ class ReducedEGProducer : public edm::stream::EDProducer<> {
   void relinkSuperCluster(T& core,
 			  const std::map<reco::SuperClusterRef, unsigned int>& superClusterMap, 
 			  const edm::OrphanHandle<reco::SuperClusterCollection>& outSuperClusterHandle);
+
```
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/interface/ReducedEGProducer.h]: One empty line is enough to separate

**Comment by arizzi** [comment]
once the review of this is done, please backport to 92X too.

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc [hunk 10]
```diff
@@ -401,6 +410,33 @@ void ReducedEGProducer::produce(edm::Event& theEvent, const edm::EventSetup& the
...
+    // Save the main gsfTrack
+    if (!gsfTrackMap.count(gsfTrack)) {
+      gsfTracks->push_back(*gsfTrack);
+      gsfTrackMap[gsfTrack] = gsfTracks->size() - 1;
+      //std::cout << "push back gsftrack to gsftrack ref" << std::endl;
+      //std::cout << " gsf track size: " << gsfTracks->size() << std::endl;
+    }
+    
+    // Save additional ambiguous gsf tracks in a map:
+    for (reco::GsfTrackRefVector::const_iterator igsf = gsfElectron.ambiguousGsfTracksBegin(); igsf != gsfElectron.ambiguousGsfTracksEnd(); ++igsf) {
```
Comment (andreypz) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: I don't want to change the loops in this PR. I based my code on the what was already there. If this must be changed, it's better to do this in a separate PR for the whole code, not just the part I committed.

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc (hunk 8 above)
Comment (andreypz) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: I do not know how to do that for these objects (ambiguous trakcs).
If you know how - please do that. Otherwise I think it's fine as it is.

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc (hunk 4 above)
Comment (andreypz) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: Same as above: if you know how, could please do that yourself. 
Otherwise I think it's fine as it is.

**Comment by andreypz** [comment]
Hi,
I implemented most of the comments from @perrotta  Please have a look.
I still left a few comments in the code, those are useful to have.

I don't know how range-based loops work (not a C++ expert here). If those must be changed, could you please do that yourself.

Thanks,
Andrey

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc (hunk 8 above)
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: Ok here: I did not realize you are looping over the ambiguousGsfTracks of the gsfElectron

File: RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc (hunk 4 above)
Comment (perrotta) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: Here you are looping over a std::vector<> , and therefore range based loops must be used:
(e.g. http://en.cppreference.com/w/cpp/language/range-for)

Comment (andreypz) [inline RecoEgamma/EgammaPhotonProducers/src/ReducedEGProducer.cc]: Ok, done.

**Comment by perrotta** [comment]
please test

**Comment by perrotta** [comment]
This PR has some effect on slimmed MET:

image

Is it expected ? Are there different/additional converted photons produced n miniAOD (but they don't get spotted by jenkins tests, if so)?Could you please comment?

**Comment by perrotta** [comment]
Running on 100 evts in TTbar13TeV2017 wf 10024.0, the new object adds 630 b/evt on miniAOD output, i.e. 1% of the total miniAOD size (63 kb/evt):

      0.0 ->       632.7        633     NEWO   1.00     recoGsfTracks_reducedEgamma_reducedGsfTracks_RECO.
    63004 ->       63419        415             0.7     ALL BRANCHES

**Comment by gpetruc** [comment]
Hi @perrotta,
the PR shouldn't change the MET. However, I see in the cmsbuild message
> The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic:
> dd94c88
and that thing changes the HF cleaning, which may affect the MET

Is there a way to re-run the test in cmsbuild? or, what was the workflow where you spotted the difference?

Giovanni

**Comment by perrotta** [comment]
Thank you @gpetruc 

The plot above with 100 evt was produced by me by comparing 100 evts in TTbar13TeV2017 wf 10024.0 in CMSSW_9_3_X_2017-08-07-1100 and CMSSW_9_3_X_2017-08-07-1100 + this PR, therefore without the additional commit which entered the jenkins tests

**Comment by perrotta** [comment]
By the way, as in the jenkins tests, also in my own apparently the only difference is in the MET significance, and not in the MET itself

**Comment by slava77** [comment]
On 8/8/17 3:05 AM, perrotta wrote:
> By the way, as in the jenkins tests, also in my own apparently the only
> difference is in the MET significance, and not in the MET itself
> 

was the extra merge added in the local tests?

We better understand where this difference in MET is coming from.

**Comment by perrotta** [comment]
Slava Krutelyov <notifications@github.com> ha scritto:

> On 8/8/17 3:05 AM, perrotta wrote:
>> By the way, as in the jenkins tests, also in my own apparently the only
>> difference is in the MET significance, and not in the MET itself
>>
>
> was the extra merge added in the local tests?

No, it was not:  
https://github.com/cms-sw/cmssw/pull/19964#issuecomment-320908506

> We better understand where this difference in MET is coming from.
>
>> —

**Comment by andreypz** [comment]
Hi @perrotta

I have two comments:

* I think your calculation of the event size increase is an overestimation.
Before this PR change, the main gsf track of an electron was "embedded" in
the electron object, now it's saved in the new collection. Therefore, the
size of the new branch is not the size increase, since large part of it was
already saved in the event before. (What's added are the additional gsf
tracks.) The increase should be calculated with and without the PR. This is
what I did at the top of this PR and it was much smaller.

* I agree that the Met should not change due to this PR. My guess is that
it is something to do with that change of "embedded" track, which now goes
into separate collection. I'm not sure how that works out exactly, and if
met calculation uses those tracks.
I can have a look at it next week, but could you tell how to reproduce
exactly the Met plot shown above.

Thanks,
Andrey

**Comment by perrotta** [comment]
Hi @andreypz

Andrey Pozdnyakov <notifications@github.com> ha scritto:

> Hi @perrotta
>
> I have two comments:
>
> * I think your calculation of the event size increase is an overestimation.
> Before this PR change, the main gsf track of an electron was "embedded" in
> the electron object, now it's saved in the new collection. Therefore, the
> size of the new branch is not the size increase, since large part of it was
> already saved in the event before. (What's added are the additional gsf
> tracks.) The increase should be calculated with and without the PR. This is
> what I did at the top of this PR and it was much smaller.

The comment in  
https://github.com/cms-sw/cmssw/pull/19964#issuecomment-320898154  
already contains all info: running wf 10024 with and without this PR  
the branch recoGsfTracks_reducedEgamma_reducedGsfTracks_RECO adds 633  
bytes/evt, and the overall miniAOD event size increases by 415  
bytes/evt, which is in line with the reduced electron size:
    4133.4 ->      3911.7       -222     -5.5  -0.35      
patElectrons_slimmedElectrons__RECO.

> * I agree that the Met should not change due to this PR. My guess is that
> it is something to do with that change of "embedded" track, which now goes
> into separate collection. I'm not sure how that works out exactly, and if
> met calculation uses those tracks.
> I can have a look at it next week, but could you tell how to reproduce
> exactly the Met plot shown above.

If you want to reproduce exactly my plot, you can run

runTheMatrix.py -l 10024.0 --command "-n 100"

in an original CMSSW_9_3_X IB (I did in CMSSW_9_3_X_2017-08-07-1100),  
and in the same IB + PR 19964, and compare the outputs of the  
step3_inMINIAODISM.root's. (But, as far as I can see from the jenkins  
tests, any other similar workflow may give the same results.)

In the comparison, you can notice that the components of the MET  
vector do not change with and without this PR: the only difference is  
in the MET significance, as reported in  
https://github.com/cms-sw/cmssw/pull/19964#issuecomment-320689166

That difference is not large, and the overall shape of the  
distributions do not change significantly. However, if this PR was not  
supposed to modify the MET, while the produced outputs show that it  
does touch it, instead, then one has to understand why.


> Thanks,
> Andrey
>

**Comment by arizzi** [comment]
is the size measured on 100 bytes? 
one should rather measure on 10k events from ttbar relval with pu

**Comment by arizzi** [comment]
I mean 100 events

**Comment by gpetruc** [comment]
I'm debugging the MET issue.

**Comment by gpetruc** [comment]
traced down to https://github.com/cms-sw/cmssw/blob/master/PhysicsTools/PatAlgos/plugins/PATElectronProducer.cc#L572-L586 which does a matching of electrons to PFCandidates by gsf track reference. will think of a fix after lunch, as it's not 100% trivial.

**Comment by gpetruc** [comment]
this should fix it https://github.com/andreypz/cmssw/compare/reducedGsfTracks-93X...gpetruc:fix_gsf_relink

**Comment by perrotta** [comment]
Thank you @gpetruc !
@andreypz : could you please (evaluate and) cherry pick the commit provided by Giovanni in this PR?

**Comment by gpetruc** [comment]
I think @andreypz may be away this week. Next week should still be in time for integration 94X, otherwise I can open a new PR from his branch + my fix.

**Comment by andreypz** [comment]
I am actually testing it right now (just running with your fix).
I made pull request to my branch here: https://github.com/andreypz/cmssw/pull/1 
I guess, once it's merged there it should propagate here.

**Comment by andreypz** [comment]
Ok, I merged a fix from @gpetruc 
I only tested it by running *test/patMiniAOD_standard_cfg.py* to make sure it still works -- it does.
I have not run the "matrix" to check met significance (it takes too long to run, and I am not at work at the moment).

**Comment by perrotta** [comment]
please test

**Comment by andreypz** [comment]
@perrotta  
I see that our bot this time suggested to do code optimization:
`curl https://cmssdt.cern.ch/SDT/code-checks/PR-19964/125/git-diff.patch | patch -p1`
should i do that as well, or is it not important?

**Comment by perrotta** [comment]
+1
- reduced gsf tracks collection were added to miniAOD as inteded
- the size increase of the miniAOD output was evaluated about 0.7% of their total size when running on 100 high PU TTbar events; as discussed above in this thread, some more precise estimate can be obtained if running on somewhat larger samples
- after the last patch, there are no more spurious effects propagated to the MET significance

**Comment by perrotta** [comment]
@andreypz : they are all valuable suggestions, but I think you can  
postpone fixing them in a separate PR, unless some update to this code  
is requested in the following of this review (and in that case: yes,  
please apply the suggested fixes, at least in the part of the code  
directly touched by your PR).


Andrey Pozdnyakov <notifications@github.com> ha scritto:

File: PhysicsTools/PatAlgos/plugins/PATElectronProducer.cc [hunk 11]
```diff
@@ -569,7 +578,15 @@ void PATElectronProducer::produce(edm::Event & iEvent, const edm::EventSetup & i
       // Is this GsfElectron also identified as an e- in the particle flow?
       bool pfId = false;
 
-      if ( pfCandsPresent ) {
+      if (usePfCandidateMultiMap_) {
+        for (const reco::PFCandidateRef pf : (*ValMultiMapH)[elePtr]) {
```
Comment (davidlange6) [inline PhysicsTools/PatAlgos/plugins/PATElectronProducer.cc]: hi @andreypz - please make this a const &

Comment (andreypz) [inline PhysicsTools/PatAlgos/plugins/PATElectronProducer.cc]: Do you mean changing to this: `const reco::PFCandidateRef& pf` ?

Comment (andreypz) [inline PhysicsTools/PatAlgos/plugins/PATElectronProducer.cc]: OK, I did it

**Comment by andreypz** [comment]
* Addressed request from @davidlange6 
* Implemented a few patches recommended by the bot. Only doing it to the files that are touched by this PR.

**Comment by andreypz** [comment]
@cmsbuild please test

**Comment by perrotta** [comment]
please test

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/19964	257a2d8
- technical update since the last reco signoff https://github.com/cms-sw/cmssw/pull/19964#issuecomment-322699583 to address review comments from @davidlange6  and also include the updates proposed by codecheck 
- jenkins tests pass

Plots for basic monitoring of the reducedGsfTracks are added in https://github.com/cms-sw/cms-bot/pull/890

**Comment by andreypz** [comment]
Hi @smuzaffar 
Could you comment on what is that code-checks-rejected about? Do you suggest to apply all the checks suggested by the bot?
My impression was that it's not mandatory, and so I only applied those related to the code I touched in my PR.

**Comment by smuzaffar** [comment]
code-checks

**Comment by smuzaffar** [comment]
@andreypz , this is new label based on the state of code-checks. Previously bot was running these checks on all files changed with respect to cmssw IB (includign those PRs which were already merged for next IB). 

Anyway, i have fixed cms-bot to run these checks for files directly changed in the PR (and any header included via those changed files) e.g. see https://github.com/cms-sw/cmssw/pull/19964#issuecomment-324005144

I would suggest to apply the changes suggested here https://cmssdt.cern.ch/SDT/code-checks/PR-19964/243/git-diff.patch

**Comment by slava77** [comment]
> Anyway, i have fixed cms-bot to run these checks for files directly changed in the PR (and any header included via those changed files)

Why does this have to extend to the header files not touched by the PR?
I consider this not mandatory, but that's just me.

**Comment by smuzaffar** [comment]
@slava77 , that was the decision we made when we enabled the code-checks.
code-checks are run for files which are in compile_commands.json and currently we only have source files (.cc, .c, .cpp etc.). So enabling header-filter option was a way to get code-checks run for header files too.

**Comment by arizzi** [comment]
I do not think that if one changes one line in one header file then we cannot accept his/her PR until the whole .h file is fixed.

**Comment by smuzaffar** [comment]
@arizzi, I agree but as we are in transition phase (non-formated code to formated code) and asking for fixing the whole file is part of this transition. At some point we will run code-checks on whole cmssw so that small changes will not fail code-checks test

**Comment by arizzi** [comment]
I think if we want people to contribute to the CMSSW code we should avoid this kind of barriers, otherwise we keep being the same ~100 people writing code for CMS.

**Comment by slava77** [comment]
On 8/22/17 5:19 AM, Malik Shahzad Muzaffar wrote:
> @slava77 <https://github.com/slava77> , that was the decision we made
> when we enabled the code-checks.
> code-checks are run for files which are in compile_commands.json and
> currently we only have source files (.cc, .c, .cpp etc.). So enabling
> header-filter option was a way to get code-checks run for header files too.

I added a comment/request to discuss this in the ORP today

**Comment by smuzaffar** [comment]
@slava77 and @arizzi  , by the way, currently code-checks does not block the PR integration. The code-check-* labels are there to show the state of code-checks. So if all L2 signs it then it will be marked as fully signed PR and ready to be integrated.

**Comment by gpetruc** [comment]
ok, I think it makes sense to not include in the PR changes to unrelated
code.

once this and #19048 are merged, I can volunteer for submitting a single PR
applying all code-checks on the PatAlgos, PatUtils and PatCandidates
packages

Giovanni

**Comment by andreypz** [comment]
Hi,
As discussed in this thread, I'm not going to apply that last code-check suggestion in this PR.
So, could you please sign it as is. Whos signature is missing here, is it @davidlange6 ?

Thanks,
Andrey

**Comment by smuzaffar** [comment]
code-checks
@slava77 , now code-checks should only apply changes to the files modified by the PR.

**Comment by smuzaffar** [comment]
looks like it did not work (still suggest changes for PhysicsTools/PatAlgos/interface/PATUserDataMerger.h).

**Comment by smuzaffar** [comment]
code-checks
another try

**Comment by davidlange6** [comment]
merge
